Dev integrate stripe

trevorhuis-smoothstack opened the pull request.

trevorhuis-smoothstack committed:
Create a new booking

trevorhuis-smoothstack committed:
Merge branch 'dev' of https://github.com/trevorhuis-smoothstack/utopia-agent into createABooking

trevorhuis-smoothstack committed:
able to cancel a booking with test

trevorhuis-smoothstack committed:
make path plural

trevorhuis-smoothstack committed:
Merge branch 'cancelAbooking' of https://github.com/trevorhuis-smoothstack/utopia-agent into devIntegrateStripe

trevorhuis-smoothstack committed:
Stripe is done but rough

trevorhuis-smoothstack committed:
updated after checking PRs to fix Req changes

trevorhuis-smoothstack committed:
remove space

trevorhuis-smoothstack committed:
I added the throwables seat flight check

trevorhuis-smoothstack commented:
This branch encompasses all of my booking and cancelling methods

justin-obrien-smoothstack commented on .gitignore:
```diff
@@ -34,4 +34,6 @@ build/
\ No newline at end of file
+/src/test/resources/application.properties
+
+src\main\java\com\ss\training\utopia\agent\Globals.java
```
Why ignore these files?

trevorhuis-smoothstack replied:
Globals is from an older solution. I'll remove it. I guess the test one doesn't matter.

justin-obrien-smoothstack commented on src/main/java/com/ss/training/utopia/agent/service/AgentService.java:
```diff
@@ -1,29 +1,160 @@
+          } catch (Exception e) {
+            return "Internal Server Error";
+            } catch (Throwable t) {
+                return "Internal Server Error";
```
You could get rid of the Exception catch block, since all Exceptions are also Throwables.

justin-obrien-smoothstack commented on src/main/java/com/ss/training/utopia/agent/service/AgentService.java:
```diff
@@ -1,29 +1,160 @@
+            );
+
+        Refund refund = Refund.create(params);
+        return refund;
```
It seems like this function doesn't need to return anything (unless having a return value makes it more testable).

justin-obrien-smoothstack commented on src/main/java/com/ss/training/utopia/agent/controller/AgentController.java:
```diff
@@ -8,32 +8,77 @@
+			case("Card Declined"):
+				break;
+			case("Flight Full"):
+				status = HttpStatus.NO_CONTENT;
```
If you're going to return the no content status, it's probably also best to send null in the body instead of a booking object. Or maybe a different status would be more fitting.

trevorhuis-smoothstack replied:
I had a lot of trouble picking this status out. I'll add the null body.

justin-obrien-smoothstack commented on src/main/java/com/ss/training/utopia/agent/controller/AgentController.java:
```diff
@@ -8,32 +8,77 @@
+				break;
+			case("Refund Processed"):
+				status = HttpStatus.OK;
+				booking.setActive(false);
```
This isn't needed here since the service has already done it (pass by reference).

trevorhuis-smoothstack replied:
You're right. Good stuff.

justin-obrien-smoothstack commented on src/test/java/com/ss/training/utopia/agent/ServiceTestSuite.java:
```diff
@@ -19,12 +24,40 @@
+        Long now = Instant.now().toEpochMilli();
+        Timestamp future = new Timestamp(now + HOUR);
+
+        // Future flight and booking - Booker ID: 1, only cancellable flight by user 1
```
This comment seems slightly off-topic for this test.

justin-obrien-smoothstack reviewed (changes requested):
I just realized that you forgot to change the number of seats available when creating and cancelling bookings. Also, the cancel booking method should use the Transactional annotation (which create booking already has).

trevorhuis-smoothstack committed:
Made changes Justin requested

justin-obrien-smoothstack reviewed (changes requested):
Cancel booking and create booking still aren't changing the number of available seats.

Also, just yesterday I learned this: In order to do a rollback, the Transactional method needs to throw an exception (by default only a RuntimeException or Error will cause a rollback, but this can be customized). Technically there's another way to rollback too, but the docs say it should be avoided unless absolutely necessary.

So this would mean that instead of catching exceptions thrown by DAOs, our service methods should be letting them through to the controller, which should be catching them and using them to decide what kind of response to send. I revised pretty much all my service methods in light of this.

https://docs.spring.io/spring/docs/3.0.x/reference/transaction.html#transaction-declarative-rolling-back

trevorhuis-smoothstack committed:
Merge branch 'dev' of https://github.com/trevorhuis-smoothstack/utopia-agent into devIntegrateStripe

trevorhuis-smoothstack committed:
change flight seats

trevorhuis-smoothstack commented:
Hey Justin, i missed the seats thing. I was just going through the inline comments for some reason. I updated it! Instead of re-writing my controllers I just added another service function for the @Transactional method to throw too. Let me know If i understood your previous comment and wrote it correctly.

justin-obrien-smoothstack commented on src/test/java/com/ss/training/utopia/agent/ServiceTestSuite.java:
```diff
@@ -96,19 +96,53 @@ public void cancelBooking() {
+       List<Flight> flights = service.readAvailableFlights();
+
+        assertEquals(flights.size(), 2);
+    }
```
For this test it would be good to make sure the flights that get found are the right ones, and also to make sure it only finds flights with available seats.

justin-obrien-smoothstack reviewed (changes requested):
The transaction handling looks like it'll work.

While the number of seats is now getting changed, those changes aren't being saved to the database.

Changes requested:
- Save flight to database after changing number of seats

trevorhuis-smoothstack committed:
final changes justin requested
